content/5_Plat/Bin_Jump.mdx

```diff
@@ -353,6 +353,76 @@ int main() {
 
 <IncompleteSection />
 
+### Solution (Company Queries II)
+
+First, let $\texttt{st}, \texttt{en}, \texttt{up}$ be the time-in, time-out, and ancestor table for all nodes in the tree. 
+
+These can be filled with a DFS traversal. $\texttt{up}$ can be generated because depth first allows the ancestors to be filled first before traversing the current node.
+
+With this information, we can declare node $a$ an ancestor of $b$ if $\texttt{st}[a] <= \texttt{st}[b] && \texttt{en}[a] >= \texttt{en}[b]$
+
+We know that if $a$ is an ancestor of $b$ or $b$ is an ancestor of $a$, the answer will be $a$ and $b$ respectively. Otherwise, we can loop from the largest jumps to the the direct ancestor of $a$ to find the LCA of $a$ and $b$. If $\texttt{up}[a][i]$ is not an ancestor of $b$, then we can set $a$ to be $\texttt{up}[a][i]$, else, we will decrement $i$ and try to find a lower common ancestor. Afterwards, our answer is not $a$, but the parent of $a$. 
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+CodeSnip{Benq Template}
+    
+int N, Q, T = 1;
+vi st, en;
+int up[200005][20];
+vi adj[200005];
+
+void dfs(int v, int p) {
+    st[v] = T++;
+    up[v][0] = p;
+
+    FOR(i, 1, 20) {
+        up[v][i] = up[up[v][i-1]][i-1];
+    }
+
+    trav(x, adj[v]) if(x != p) dfs(x, v);
```

Comment:
don't benq template use `each` instead of `trav`?

Reply:
:O am i using outdated benq template

Reply:
OMG its each now... :(